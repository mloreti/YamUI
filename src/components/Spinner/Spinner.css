/*! Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license. */
@import 'src/css/variables/colors.css';
@import 'src/css/variables/fonts.css';

.y-spinner {
  display: flex;
  align-items: center;
  flex-direction: row;
}
.y-spinner__centered {
  justify-content: center;
}

.y-spinner--circle > .ms-Spinner-circle {
  border: 2px solid color($spinnerColor__light a(30%));
  border-top-color: $spinnerColor__light;
}

.y-spinner__color-dark > .y-spinner--circle > .ms-Spinner-circle {
  border: 2px solid color($spinnerColor__dark a(30%));
  border-top-color: $spinnerColor__dark;
}

.y-spinner--label {
  margin: 0 0 0 8px;
}

.y-spinner__size-xSmall > .y-spinner--label {
  margin: -1px 0 0 6px;
}

.y-spinner__size-small > .y-spinner--label {
  margin: -1px 0 0 8px;
}

.y-spinner__size-medium > .y-spinner--label {
  margin: -2px 0 0 8px;
}

.y-spinner__size-large {
  flex-direction: column;
}

.y-spinner__size-large > .y-spinner--label {
  margin: 4px 0 0;
}
