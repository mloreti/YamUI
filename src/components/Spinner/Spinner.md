### Examples

With visible text:

```js { "props": { "data-description": "with visible text" } }
const { SpinnerColor, SpinnerSize } = require('.');

<table style={{ textAlign: 'center' }}>
  <thead>
    <tr>
      <th style={{ width: '100px' }}>L</th>
      <th style={{ width: '100px' }}>M</th>
      <th style={{ width: '100px' }}>S</th>
      <th style={{ width: '100px' }}>XS</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>
        <Spinner size={SpinnerSize.LARGE} text="Loading" isCentered={true} />
      </td>
      <td>
        <Spinner size={SpinnerSize.MEDIUM} text="Loading" isCentered={true} />
      </td>
      <td>
        <Spinner size={SpinnerSize.SMALL} text="Loading" isCentered={true} />
      </td>
      <td>
        <Spinner size={SpinnerSize.XSMALL} text="Loading" isCentered={true} />
      </td>
    </tr>
    <tr>
      <td>
        <Spinner size={SpinnerSize.LARGE} text="Loading" isCentered={true} color={SpinnerColor.METADATA} />
      </td>
      <td>
        <Spinner size={SpinnerSize.MEDIUM} text="Loading" isCentered={true} color={SpinnerColor.METADATA} />
      </td>
      <td>
        <Spinner size={SpinnerSize.SMALL} text="Loading" isCentered={true} color={SpinnerColor.METADATA} />
      </td>
      <td>
        <Spinner size={SpinnerSize.XSMALL} text="Loading" isCentered={true} color={SpinnerColor.METADATA} />
      </td>
    </tr>
    <tr style={{ backgroundColor: '#386cbb' }}>
      <td>
        <Spinner color={SpinnerColor.DARK} size={SpinnerSize.LARGE} text="Loading" isCentered={true} />
      </td>
      <td>
        <Spinner color={SpinnerColor.DARK} size={SpinnerSize.MEDIUM} text="Loading" isCentered={true} />
      </td>
      <td>
        <Spinner color={SpinnerColor.DARK} size={SpinnerSize.SMALL} text="Loading" isCentered={true} />
      </td>
      <td>
        <Spinner color={SpinnerColor.DARK} size={SpinnerSize.XSMALL} text="Loading" isCentered={true} />
      </td>
    </tr>
  </tbody>
</table>
```

With invisible text:

```js { "props": { "data-description": "with invisible text" } }
const { SpinnerColor, SpinnerSize } = require('.');

<table style={{ textAlign: 'center' }}>
  <thead>
    <tr>
      <th style={{ width: '100px' }}>L</th>
      <th style={{ width: '100px' }}>M</th>
      <th style={{ width: '100px' }}>S</th>
      <th style={{ width: '100px' }}>XS</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>
        <Spinner size={SpinnerSize.LARGE} text="Loading" hideText={true} isCentered={true} />
      </td>
      <td>
        <Spinner size={SpinnerSize.MEDIUM} text="Loading" hideText={true} isCentered={true} />
      </td>
      <td>
        <Spinner size={SpinnerSize.SMALL} text="Loading" hideText={true} isCentered={true} />
      </td>
      <td>
        <Spinner size={SpinnerSize.XSMALL} text="Loading" hideText={true} isCentered={true} />
      </td>
    </tr>
    <tr style={{ backgroundColor: '#386cbb' }}>
      <td>
        <Spinner color={SpinnerColor.DARK} size={SpinnerSize.LARGE} text="Loading" hideText={true} isCentered={true} />
      </td>
      <td>
        <Spinner color={SpinnerColor.DARK} size={SpinnerSize.MEDIUM} text="Loading" hideText={true} isCentered={true} />
      </td>
      <td>
        <Spinner color={SpinnerColor.DARK} size={SpinnerSize.SMALL} text="Loading" hideText={true} isCentered={true} />
      </td>
      <td>
        <Spinner color={SpinnerColor.DARK} size={SpinnerSize.XSMALL} text="Loading" hideText={true} isCentered={true} />
      </td>
    </tr>
  </tbody>
</table>
```

With isCentered:

```js { "props": { "data-example": "centered" } }

<div>
  <div>
    <p>Default alignment:</p>
    <Spinner text="Loading" />
  </div>
  <br />
  <div>
    <p>Centered within its container:</p>
    <Spinner text="Loading" isCentered={true} />
  </div>
</div>