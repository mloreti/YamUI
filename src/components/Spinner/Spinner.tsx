/*! Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license. */
import '../../yamui';
import * as React from 'react';
import {
  ISpinnerProps as FabricSpinnerProps,
  Spinner as FabricSpinner,
  SpinnerSize as FabricSpinnerSize,
} from 'office-ui-fabric-react/lib/Spinner';
import { BaseComponentProps } from '../../util/BaseComponent/props';
import Text, { TextProps, TextColor, TextSize } from '../Text';
import ScreenreaderText from '../ScreenreaderText';
import { SpinnerSize } from './enums';
import './Spinner.css';

export { SpinnerSize };

const SizeMap = {
  [SpinnerSize.LARGE]: FabricSpinnerSize.large,
  [SpinnerSize.MEDIUM]: FabricSpinnerSize.medium,
  [SpinnerSize.SMALL]: FabricSpinnerSize.small,
  [SpinnerSize.XSMALL]: FabricSpinnerSize.xSmall,
};

export enum SpinnerColor {
  /**
   * Light theme.
   */
  LIGHT = 'light',
  /**
   * Light theme spinner color with gray text.
   */
  METADATA = 'metadata',
  /**
   * Dark theme.
   */
  DARK = 'dark',
}

export interface SpinnerProps extends BaseComponentProps {
  /**
   * Text displayed besides/below the spinner.
   */
  text: string;

  /**
   * Whether to hide text, and only render it for screenreaders.
   * @default false
   */
  hideText?: boolean;

  /**
   * Color of the spinner.
   * @default SpinnerColor.LIGHT
   */
  color?: SpinnerColor;

  /**
   * Size of the spinner.
   * @default SpinnerSize.MEDIUM
   */
  size?: SpinnerSize;

  /**
   * Will center the Spinner horizontally within its container.
   */
  isCentered?: boolean;
}

/**
 * A `Spinner` is an outline of a circle which animates around itself indicating to the user that
 * things are processing. It is shown when we're unsure how long a task will take.
 */
export default class Spinner extends React.Component<SpinnerProps> {
  public static defaultProps: Partial<SpinnerProps> = {
    hideText: false,
    color: SpinnerColor.LIGHT,
    size: SpinnerSize.MEDIUM,
  };

  public render() {
    const { text, hideText } = this.props;

    const textComponent = hideText ? (
      <ScreenreaderText>{text}</ScreenreaderText>
    ) : (
      <Text {...this.getLabelProps()}>{text}</Text>
    );

    return (
      <div className={this.getClasses()}>
        <FabricSpinner {...this.getSpinnerProps()} />
        {textComponent}
      </div>
    );
  }

  private getClasses() {
    const { isCentered, className, color, size } = this.props;

    const classes: string[] = ['y-spinner', `y-spinner__color-${color}`, `y-spinner__size-${size}`];

    if (isCentered) {
      classes.push('y-spinner__isCentered');
    }

    if (className) {
      classes.push(className);
    }

    return classes.join(' ');
  }

  private getSpinnerProps(): FabricSpinnerProps {
    const { size } = this.props;

    return {
      className: 'y-spinner--circle',
      size: SizeMap[size as SpinnerSize],
    };
  }

  private getLabelProps(): TextProps {
    const { color, size } = this.props;
    let textColor = TextColor.PRIMARY;
    if (color === SpinnerColor.DARK) {
      textColor = TextColor.WHITE;
    } else if (color === SpinnerColor.METADATA) {
      textColor = TextColor.METADATA;
    }

    return {
      color: textColor,
      size: size === SpinnerSize.XSMALL ? TextSize.SMALL : TextSize.MEDIUM_SUB,
      className: 'y-spinner--label',
    };
  }
}
