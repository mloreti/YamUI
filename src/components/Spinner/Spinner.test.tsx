/*! Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license. */
import * as React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';
import Spinner, { SpinnerProps, SpinnerColor, SpinnerSize } from '.';

describe('<Spinner />', () => {
  let component: ShallowWrapper<SpinnerProps>;

  describe('with additional className', () => {
    beforeEach(() => {
      component = shallow(<Spinner className="TEST_CLASSNAME" text="Loading" />);
    });

    it('matches its snapshot', () => {
      expect(component).toMatchSnapshot();
    });
  });

  describe('with visible text', () => {
    beforeEach(() => {
      component = shallow(<Spinner text="Loading" />);
    });

    it('matches its snapshot', () => {
      expect(component).toMatchSnapshot();
    });
  });

  describe('with invisible text', () => {
    beforeEach(() => {
      component = shallow(<Spinner text="Loading" hideText={true} />);
    });

    it('matches its snapshot', () => {
      expect(component).toMatchSnapshot();
    });
  });

  describe('with dark color', () => {
    beforeEach(() => {
      component = shallow(<Spinner text="Loading" color={SpinnerColor.DARK} />);
    });

    it('matches its snapshot', () => {
      expect(component).toMatchSnapshot();
    });
  });

  describe('with metadata color', () => {
    beforeEach(() => {
      component = shallow(<Spinner text="Loading" color={SpinnerColor.METADATA} />);
    });

    it('matches its snapshot', () => {
      expect(component).toMatchSnapshot();
    });
  });

  describe('with a size', () => {
    beforeEach(() => {
      component = shallow(<Spinner text="Loading" size={SpinnerSize.XSMALL} />);
    });

    it('matches its snapshot', () => {
      expect(component).toMatchSnapshot();
    });
  });

  describe('when centered', () => {
    beforeEach(() => {
      component = shallow(<Spinner text="Loading" isCentered={true} />);
    });

    it('matches its snapshot', () => {
      expect(component).toMatchSnapshot();
    });
  });
});
