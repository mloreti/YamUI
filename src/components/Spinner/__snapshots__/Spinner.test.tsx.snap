// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Spinner /> when centered matches its snapshot 1`] = `
<div
  className="y-spinner y-spinner__color-light y-spinner__size-medium y-spinner__isCentered"
>
  <Spinner
    ariaLive="polite"
    className="y-spinner--circle"
    size={2}
  />
  <Text
    className="y-spinner--label"
    color="primary"
    size="mediumSub"
  >
    Loading
  </Text>
</div>
`;

exports[`<Spinner /> with a size matches its snapshot 1`] = `
<div
  className="y-spinner y-spinner__color-light y-spinner__size-xSmall"
>
  <Spinner
    ariaLive="polite"
    className="y-spinner--circle"
    size={0}
  />
  <Text
    className="y-spinner--label"
    color="primary"
    size="small"
  >
    Loading
  </Text>
</div>
`;

exports[`<Spinner /> with additional className matches its snapshot 1`] = `
<div
  className="y-spinner y-spinner__color-light y-spinner__size-medium TEST_CLASSNAME"
>
  <Spinner
    ariaLive="polite"
    className="y-spinner--circle"
    size={2}
  />
  <Text
    className="y-spinner--label"
    color="primary"
    size="mediumSub"
  >
    Loading
  </Text>
</div>
`;

exports[`<Spinner /> with dark color matches its snapshot 1`] = `
<div
  className="y-spinner y-spinner__color-dark y-spinner__size-medium"
>
  <Spinner
    ariaLive="polite"
    className="y-spinner--circle"
    size={2}
  />
  <Text
    className="y-spinner--label"
    color="white"
    size="mediumSub"
  >
    Loading
  </Text>
</div>
`;

exports[`<Spinner /> with invisible text matches its snapshot 1`] = `
<div
  className="y-spinner y-spinner__color-light y-spinner__size-medium"
>
  <Spinner
    ariaLive="polite"
    className="y-spinner--circle"
    size={2}
  />
  <ScreenreaderText>
    Loading
  </ScreenreaderText>
</div>
`;

exports[`<Spinner /> with metadata color matches its snapshot 1`] = `
<div
  className="y-spinner y-spinner__color-metadata y-spinner__size-medium"
>
  <Spinner
    ariaLive="polite"
    className="y-spinner--circle"
    size={2}
  />
  <Text
    className="y-spinner--label"
    color="metadata"
    size="mediumSub"
  >
    Loading
  </Text>
</div>
`;

exports[`<Spinner /> with visible text matches its snapshot 1`] = `
<div
  className="y-spinner y-spinner__color-light y-spinner__size-medium"
>
  <Spinner
    ariaLive="polite"
    className="y-spinner--circle"
    size={2}
  />
  <Text
    className="y-spinner--label"
    color="primary"
    size="mediumSub"
  >
    Loading
  </Text>
</div>
`;
